src/artifactRegistry.ts

```diff
@@ -124,3 +205,103 @@ export class CachingDockerRegistryClient {
     return tags;
   }
 }
+
+/**
+ * An example format of a docker tag:
+ * {
+ *   "name":"projects/platform-cross-environment/locations/us-central1/repositories/platform-docker/packages/identity/tags/2022.02-278-g123456789",
+ *   "version":"projects/platform-cross-environment/locations/us-central1/repositories/platform-docker/packages/identity/versions/sha256:123abc456defg"
+ * }
+ *
+ */
+export type DockerTag = {
+  name: string;
+  version: string;
+};
+
+/**
+ * getRelevantCommits
+ *
+ * @param {string} prevTag
+ *   The previous docker tag: of the following format: main---0013567-2024.04-g<githash>
+ * @param {string} nextTag
+ *  The next docker tag: main---0013567-2024.04-g<githash>
+ *
+ * @param {DockerTag[]} dockerTags
+ *  The docker tags as we need to parse and filter into relevant commits. Provided by a previous call to the artifact registry.
+ *
+ *
+ * @param {function} getTagFromDockerTag
+ *   Should map from `projects/platform-cross-environment/locations/us-central1/repositories/platform-docker/packages/servicename/tags/2022.02-278-g123456789` -> `2022.02-278-g123456789`
+ *   Should be a wrapper around the ArtifactRegistryClient, but written this way for testability, so the behavior can be injected.
+ *
+ * @returns {string[]} - The relevant commits as strings
+ */
+export function getRelevantCommits(
+  prevTag: string,
+  nextTag: string,
+  dockerTags: DockerTag[],
+  getTagFromDockerTag: (dockerTag: DockerTag) => string,
+): string[] {
+  if (!prevTag.startsWith('main')) return [];
+  if (!nextTag.startsWith('main')) return [];
+
+  /**
+   * Going to loop over our docker tags, filtering out ones outside the relevant range, and build an array of tag info.
+   *
+   * We will then do some commit deduping and return a list of relevant commits.
+   */
+  const relevantCommitsWithTagInfo = new Array<{
+    version: string;
+    tag: string;
+    commit: string;
+  }>();
+
+  for (const dockerTag of dockerTags) {
+    if (!dockerTag.version || !dockerTag.name) continue;
+
+    const tag = getTagFromDockerTag(dockerTag);
+
+    if (!tag.startsWith('main')) continue;
+    if (!tagInRange(prevTag, nextTag, tag)) continue;
+
+    // We only care about the tags between prev and next that have a git commit
+    const gitCommitMatches = tag.match(/-g([0-9a-fA-F]+)$/);
```

Comment:
I'd be tempted to say that this "ArtifactRegistry"-level API should just be trying to return Docker tags that demonstrate an underlying image version change, and not caring about whether they contain git commits in them or not.  Then a layer up can take the list of changed docker tags and map them into git commits.  Feels like the right layering of concerns to me unless there's an important performance reason not to.